app/main/helpers/frameworks.py

```diff
@@ -365,7 +364,47 @@ def status_priority(status):
 
 
 def get_supplier_registered_name_from_declaration(declaration):
-    return(
+    return (
         declaration.get('supplierRegisteredName')  # G-Cloud 10 and later declaration key
         or declaration.get('nameOfOrganisation')  # G-Cloud 9 and earlier declaration key
     )
+
+
+class EnsureApplicationCompanyDetailsHaveBeenConfirmed:
+    """A decorator for framework application views that should not be accessible before company details have
+    been confirmed for that application.
+
+    This is a class-based decorator primarily so that it's possible to mock out the validator for the majority of tests
+    which aren't actively testing that the view requires suppliers to be in a certain application state."""
+
+    def __init__(self, data_api_client):
+        self.data_api_client = data_api_client
+
+    def __call__(self, func):
+        @wraps(func)
+        def decorated_view(*args, **kwargs):
+            if self.validator(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            # Shouldn't be able to get here
+            abort(500, "There was a problem accessing this page of your application. Please try again later.")
+
+        return decorated_view
+
+    def validator(self, *args, **kwargs):
+        """Performs the actual validation that ensures the logged-in supplier has confirmed their company details
+        are correct for this application"""
+        if 'framework_slug' not in kwargs:
+            current_app.logger.error("Required parameter `framework_slug` is undefined for the calling view.")
+            abort(500, "There was a problem accessing this page of your application. Please try again later.")
+
+        if current_user.is_authenticated() and current_user.supplier_id:
```

Comment:
Do we need to check they're authenticated? If they aren't then the validation will pass anyway? And I'm guessing this needs to be used on routes that have the `@login_required` decorator?

Reply:
Hmm, maybe we don't need to check if they're authenticated. I didn't really check how the user object works - I thought we might not be able to access supplier_id without an authenticated users, and the decorator theoretically could be called before login_required.

Reply:
Yeah, I just checked and supplier_id throws an attribute error on an anonymous user. so the `is_authenticated` check is necessary in case this decorator is called before `login_required`.